### ISSUE TITLE
Inquiry Regarding the Hyperparameter “unnorm_key”

### ISSUE THREAD

**Issue by MstarLioning** [SpatialVLA/SpatialVLA#24]
Hello, sorry to bother you again. As a beginner in VLA, I am curious about the hyperparameter “unnorm_key” in SpatialVLA and its role (mapping it back to the real space through the formula 0.5 * (normalized_actions + 1) * (action_high - action_low) + action_low).

My main confusion is why a single model can have different unnorm_keys for one set of parameters. In my personal thought (or guess), the training process of VLA, in addition to mimicking the trajectory, also requires it to “understand” the ratio between its output values and the actual spatial displacement and rotation. Therefore, my understanding is that one set of parameters should only be able to learn (or comprehend) one fixed scaling ratio between the model’s outputs and the real space.

An even bigger question is why we need to explicitly state the unnorm_key. After examining the processor_config.json file of SpatialVLA, I found that the unnorm_key queries a series of statistical features, which are obtained during training. So why isn’t the mapping from the normalized data to the real space (translation and rotation) a fixed mapping relationship, and why does it change with different datasets? (In other words, what should be done if, in an OOD (out-of-distribution) scenario, a new sample arrives for prediction that cannot be classified into any specific dataset?)

I have observed that for other VLAs (π0, oepnvla, DexGraspVLA) the unnorm is generally represented in this way. Why don’t they all use a fixed mapping pattern? Could you please share your thoughts or analysis on this hyperparameter?

I am eagerly awaiting your reply! I hope to receive your guidance!

您好，不好意思不知道能否再冒昧打扰您一下：
作为一名在VLA方面的初学者，我对于SpatialVLA中unnorm_key这个超参数以及其作用（通过0.5 * (normalized_actions + 1) * (action_high - action_low) + action_low将其映射回真实空间）比较好奇。

我主要疑惑的地方在于为什么针对一个模型的一套参数可以有不同的unnorm_key？在我个人的想法（或者猜测）里VLA训练的过程除了去模仿轨迹之外还需要让其去“理解”其输出数值与真实空间位移、旋转的比值。所以我的认知是一套参数不应该只能学习（理解）一套模型输出与真实空间缩放比例么？

而更大的疑惑在于我不是很理解为什么要显式的表述出unnorm_key？在学习SPatialVLA的processor_config.json文件后发现unnorm_key查询的对象是一系列统计特征，而这个统计特征又是在训练中得到的，那为什么从归一化数据到真实空间的平动转动不是一个固定的映射关系而是会随着数据集的变化而变化呢？（换言之如果在OOD的情况下有一个新的样本等待预测而不能将其归为某类特定的数据集这种情况该怎么办呢？）

我在横向观察其他VLA（π0、oepnvla、DexGraspVLA）的时候发现针对unnorm基本都是这样表示的，为什么大家不把他做成一个固定映射的模式呢？能否请您说一下对这个超参的看法或解析呢？

非常期待您的回复！希望能得到您的指点！

**Comment by DelinQu** [comment]
您好！
主要原因是openx不同数据集的机械臂构型很不一样，而且尺度差别非常大，如果都用统一的mean std会导致某些数据集分布在非常小或非常大的范围内，具身的数据相对于imagenet这种cv的image差别还是太大了。在训练的时候数据送到一个dataloader分发给模型训练，数据根据不同mean std和normalized 到[-1,1]的范围。推理的时候unnomalize出来。当然如果不做pretrain，只在某些小数据集上tuning这个问题会小很多。

**Comment by DelinQu** [comment]
1. 一般来说unnorm_key是针对某一个数据集的，而不是机械臂，最好是每个数据集统计一下。
2. rlds在训练前会统计mean std，对应代码是https://github.com/SpatialVLA/SpatialVLA/blob/53a59f91ad90e5b8d4eb409e9ebae311e847dd98/data/utils/data_utils.py#L244
3. fine-tuning的时候，一般是固定几个小数据集，一个数据集训一次模型，那么这个key直接存到模型里就好了，不需要维护一个很大的dict。

**Comment by DelinQu** [comment]
如果没有数据集tuning的话，可以试一试

**Comment by MstarLioning** [comment]
收到，谢谢您！我会尽快去尝试！再次感谢您耐心的解答！
